File: leetcode/CPP/56. Merge Intervals.cpp
// https://leetcode.com/problems/merge-intervals/
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> output;
        if(intervals.size() ==0) return output;
        sort(intervals.begin(), intervals.end());
        auto temp = intervals[0];
        for(int i=1;i<intervals.size();i++){
            if(temp[1]>=intervals[i][0]){
                temp[1] = max(temp[1], intervals[i][1]);
            }else{
                output.push_back(temp);
                temp=intervals[i];
            }
        }
        output.push_back(temp);
        return output;
    }
};

File: Leetcode/Java/56.Merge Intervals.java
// Time:  O(nlogn)
// Space: O(n)
import java.util.Arrays;
import java.util.LinkedList;

class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        LinkedList<int[]> result = new LinkedList<>();
        
        for (int[] interval : intervals) {
            if (result.isEmpty() || result.getLast()[1] < interval[0]) {
                result.add(interval);
            } else {
                result.getLast()[1] = Math.max(result.getLast()[1], interval[1]);
            }
        }
        
        return result.toArray(new int[result.size()][]);
    }
}
